---

title: Java 异常
date: 2020-01-13 14:16:21
tags:
- Java
typora-root-url: ..
---

#### Java 异常类层次结构

![image-20210607191130717](/post_imgs/java_2-4-1.png)

##### Throwable 类

Throwable 类是 Java 语言中所有错误或异常的超类。

- 只有当对象是此类的实例时，才能通过 Java 虚拟机或者 Java throw 语句抛出。类似地，只有此类或其子类之一才可以是 catch 子句中的参数类型。

- Throwable 包含了其线程创建时线程`执行堆栈`的快照。它还包含了给出有关错误更多信息的消息字符串。

- 最后，它还可以包含 cause（原因）：另一个导致此 throwable 抛出的 throwable。此 cause 设施在 1.4 版本中首次出现。它也称为`异常链`设施，因为 cause 自身也会有 cause，依此类推，就形成了异常链，每个异常都是由另一个异常引起的。

##### 异常分类

Throwable类有两个直接子类：
   - Exception：是程序本身可以处理的异常。Exception 类有一个重要的子类 RuntimeException。RuntimeException 类及其子类表示“JVM 常用操作”引发的错误。例如，若试图使用空值对象引用、除数为零或数组越界，则分别引发运行时异常NullPointerException、ArithmeticException 和 ArrayIndexOutOfBoundException。

   - Error：系统错误，用于指示合理的应用程序**不应该试图捕获的严重问题**。，表示运行应用程序中较严重问题，表示代码运行时 JVM 出现的问题。例如，Java虚拟机运行错误（Virtual MachineError），当 JVM 不再有继续执行操作所需的内存资源时，将出现 OutOfMemoryError。这些异常发生时，Java虚拟机（JVM）一般会选择线程终止。

  `注意`：异常和错误的区别：异常能被程序本身可以处理，错误是无法处理。

可查异常与不可查异常：

Java的异常(包括Exception和Error)分为可查的异常（checked exceptions）和不可查的异常（unchecked exceptions）

- Checked Exceptions：编译器要求必须处置的异常，可查异常虽然是异常状况，但在一定程度上它的发生是可以预计的，而且一旦发生这种异常状况，就必须采取某种方式进行处理。
- Unchecked Exceptions：编译器不要求强制处置的异常，包括运行时异常（RuntimeException与其子类）和 错误（Error）。

运行时异常与非运行时异常：

- 运行时异常：都是RuntimeException类及其子类异常，如NullPointerException、IndexOutOfBoundsException等，这些异常是不检查异常，程序中可以选择捕获处理，也可以不处理。这些异常一般是由程序逻辑错误引起的，程序应该从逻辑角度尽可能避免这类异常的发生。运行时异常的特点是Java编译器不会检查它。
- 非运行时异常 （编译异常）：是RuntimeException以外的异常，类型上都属于Exception类及其子类。从程序语法角度讲是必须进行处理的异常，如果不处理，程序就不能编译通过。如IOException、SQLException等以及用户自定义的Exception异常

##### 自定义异常

Java 的异常机制中所定义的所有异常不可能预见所有可能出现的错误，某些特定的情境下，则需要我们自定义异常类型来向上报告某些错误信息。

在 Java 中你可以自定义异常。编写自己的异常类时需要记住下面的几点。

- 所有异常都必须是 Throwable 的子类。
- 如果希望写一个检查性异常类，则需要继承 Exception 类。
- 如果你想写一个运行时异常类，那么需要继承 RuntimeException 类。

##### 异常处理机制

- try...catch...finally关键字

  - 一个 try 代码块后面跟随多个 catch 代码块的情况就叫**多重捕获**。
  - **无论是否发生异常，finally 代码块中的代码总会被执行**。在 finally 代码块中，可以运行清理类型等收尾善后性质的语句。

- throws/throw 关键字

  如果一个方法没有捕获一个检查性异常，那么该方法**必须使用 throws 关键字来声明**。throws 关键字放在方法签名的尾部。**也可以使用 throw 关键字抛出一个异常**，无论它是新实例化的还是刚捕获到的。

